### DISCUSSION TITLE
phase4-peppol-server-webapp cannot decrypt the incoming message due to X509PKIPathv1

### DISCUSSION THREAD

**Discussion by binaradarsha** [phax/phase4#149] [General]
Hi,

When I send a SBDH doc using `Phase4PeppolSender.sbdhBuilder()` to `phase4-peppol-server-webapp`, the incoming message header includes two BinarySecurityTokens and second one has `ValueType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509PKIPathv1"`. Because of that decryption fails.

Is this a client side problem that second BinarySecurityToken has `ValueType="...X509PKIPathv1"` instead of `X509v3`, or a lack of functionality in `phase4-peppol-server-webapp`?

**Comment by phax** [comment]
Hi @binaradarsha okay, interesting. Well, that is a regression for a different profile where I thought it has no impact. I will revert the change for the Peppol profile. Please stay tuned.

**Reply by binaradarsha** [reply]
Hi,
I checked sending a document from `phase4-peppol-client`, it was successful. Both BinarySecurityTokens has ValueType="...X509v3".
Second BinarySecurityToken has ValueType="...X509PKIPathv1" when only send from our client program. It has following dependencies. Could you please advise if something is missing or wrong?

  "com.helger.phase4" % "phase4-profile-bpc" % "2.1.3",
  "com.helger.phase4" % "phase4-profile-peppol" % "2.1.3",
  "com.helger.phase4" % "phase4-dynamic-discovery" % "2.1.3",
  "com.helger.phase4" % "phase4-peppol-client" % "2.1.3",
  "jakarta.servlet" % "jakarta.servlet-api" % "6.0.0",

**Reply by phax** [reply]
It is fixed in the 2.1.5 release which will be available later today

**Reply by binaradarsha** [reply]
Thanks Philip, we will check the new version. 

By the way, usually a peppol certificate is in a chain with CA cert in keystore/truststore. So which is the correct way of include it in security header, as a single cert (X509v3) or as a cert chain (X509PKIPathv1)?

**Reply by phax** [reply]
In the security header, it is indeed included as a single certificate only. The full certificate chain is only needed for signing, but not for transmission, because we have a fixed PKI in Peppol.

**Reply by phax** [reply]
Release 2.1.5 is available at https://github.com/phax/phase4/releases/tag/phase4-parent-pom-2.1.5 and soon from Maven Central

**Reply by binaradarsha** [reply]
Thank you very much Philip. We will test when the changes are available in Maven.

**Reply by phax** [reply]
They are available now

**Reply by binaradarsha** [reply]
Hi, your fix worked properly. Couldn't update early, very sorry.

**Reply by phax** [reply]
Thanks - no worries. I confirmed it myself in the meantime :)
